src/coreclr/vm/threadsuspend.cpp

```diff
@@ -1227,176 +1236,188 @@ Thread::UserAbort(EEPolicy::ThreadAbortTypes abortType, DWORD timeout)
 
-            // We need to make sure that ThreadStoreLock is released after CheckForAbort.  This makes sure
-            // that ThreadAbort does not race against GC.
-            class CheckForAbort
+        // We need to make sure that ThreadStoreLock is released after CheckForAbort.  This makes sure
+        // that ThreadAbort does not race against GC.
+        class CheckForAbort
+        {
+        private:
+            Thread *m_pThread;
+            BOOL m_fHoldingThreadStoreLock;
+            BOOL m_NeedRelease;
+        public:
+            CheckForAbort(Thread *pThread, BOOL fHoldingThreadStoreLock)
+            : m_pThread(pThread),
+            m_fHoldingThreadStoreLock(fHoldingThreadStoreLock),
+            m_NeedRelease(FALSE)
             {
-            private:
-                Thread *m_pThread;
-                BOOL m_fHoldingThreadStoreLock;
-                BOOL m_NeedRelease;
-            public:
-                CheckForAbort(Thread *pThread, BOOL fHoldingThreadStoreLock)
-                : m_pThread(pThread),
-                m_fHoldingThreadStoreLock(fHoldingThreadStoreLock),
-                m_NeedRelease(TRUE)
-                {
-                    if (!fHoldingThreadStoreLock)
-                    {
-                        ThreadSuspend::LockThreadStore(ThreadSuspend::SUSPEND_OTHER);
-                    }
-                    ThreadStore::ResetStackCrawlEvent();
-
-                    // The thread being aborted may clear the TS_AbortRequested bit and the matching increment
-                    // of g_TrapReturningThreads behind our back. Increment g_TrapReturningThreads here
-                    // to ensure that we stop for the stack crawl even if the TS_AbortRequested bit is cleared.
-                    ThreadStore::TrapReturningThreads(TRUE);
-                }
-                void NeedStackCrawl()
-                {
-                    m_pThread->SetThreadState(Thread::TS_StackCrawlNeeded);
-                }
-                ~CheckForAbort()
+            }
+            void Activate()
+            {
+                m_NeedRelease = TRUE;
+                if (!m_fHoldingThreadStoreLock)
                 {
-                    Release();
+                    ThreadSuspend::LockThreadStore(ThreadSuspend::SUSPEND_OTHER);
                 }
-                void Release()
+                ThreadStore::ResetStackCrawlEvent();
+
+                // The thread being aborted may clear the TS_AbortRequested bit and the matching increment
+                // of g_TrapReturningThreads behind our back. Increment g_TrapReturningThreads here
+                // to ensure that we stop for the stack crawl even if the TS_AbortRequested bit is cleared.
+                ThreadStore::TrapReturningThreads(TRUE);
+            }
+            void NeedStackCrawl()
+            {
+                m_pThread->SetThreadState(Thread::TS_StackCrawlNeeded);
```

Comment:
Nit: This assumes `NeedStackCrawl` is never called on a non-`Activate`d instance.  We might want to add an assert or a check here.